### ISSUE TITLE
Add *%_ to documentation for *all* method signatures 

### ISSUE THREAD

**Issue by codesections** [Raku/doc#3578]
After discussing #3576 in the #raku IRC channel, I've decided to close that issue in favor of a new approach.

## Problem

As documented at [type/Method](https://docs.raku.org/type/Method)
> Methods will ignore extra named arguments where other types of `Routine` will throw at runtime. Extra arguments will be forwarded by `nextsame` and friends.

Another way to say this is that each method declaration implicitly adds `*%_` to the method's signature – which can be confirmed by calling `.signature` on any `Method`.  Indeed, [Roast requires](https://github.com/Raku/roast/blob/master/S12-class/interface-consistency.t) that `.signature` display the `*%_` parameter.

However, _none_ of the method signatures the documentation list include this parameter.

## Suggestions

Add `*%_` to all method signatures (through a combination of automated tooling and manual review) and add a brief note to the /type/Method page clarifying that methods receive an implicit `*$_`  parameter. 

As jdv79 pointed out on IRC, this would help new users avoid confusion from looking at the signature listed in the docs and being confused by the fact that a method's signature doesn't exactly match the signature displayed in the docs.  It's possible that it could _create_ confusion for a different class of new user – one so new that they are confused by the `*$_` syntax.  Hopefully, however, adding a clear note in the Method docs will prevent that confusion (and give very new users a chance to learn about some of Raku's cooler features!)

Because this change will impact a significant number of files in `doc/Type` (er, well, _all_ of them), I won't do any work on this issue for a few days in case anyone has objections to this change.

**Comment by JJ** [comment]
Just to be clear, the suggestions do *not* look fine to me. In some cases where it's actually used (say, the infamous `lines` where those arguments are passed through) it might be helpful, but on a case by case basis. Please revert #3601 and let's talk about the best solution for this.

**Comment by codesections** [comment]
Again, my apologies for the miscommunication; I explained how it happened a bit in a comment #3601, and have updated this issue's title to clarify the scope of the proposed changes.

Since there was a miscommunication, let me back up and explain where I'm coming from with this change.

* The documentation currently lists the signature for each documented method
* Roast requires that `Method`s have a `signature` method, which also lists the signature for each method
* In an **ideal** world (one with a 100% complete Roast, a 100% accurate implementation, and a 100% accurate set of docs), these two ways of viewing the signature for the method would show the same signature
* In the real world, we'll likely never get to a point where the documented signatures and the signatures listed by Rakudo's `.signature` are exactly the same, but it's worth aiming for
* Discrepancies between the documented signature and the `.signature` output could indicate a 1) a Rakudo bug, 2) an error in the docs, or 3) an under-specified area in Roast

Given all of the above, I would _like_ to get the signatures that we document closer to the signatures returned by  `.signature`.  Doing so would allow us to check for remaining discrepancies, which could indicate issues (either for us, Roast, or Rakudo).  However, right now exactly **none** of the signatures we document are the same as the `.signature` signatures.

Consider the `Str.chomp` method for example:  If you consult [our docs](https://docs.raku.org/routine/chomp#(Str)_routine_chomp) you'll see that it has the signature 

```raku
(Str:D: --> Str:D)
```

However, if you run `Str.^lookup('chomp').signature` in a REPL, you'll get

```raku
(Str:D: *%_ --> Str:D)
```

Additionally, this output from `.signature` is _not_ a choice Rakudo made – it's [explicitly specified in Roast](https://github.com/Raku/roast/blob/master/S12-class/interface-consistency.t#L19).  And including `*%_` in the signature for `Str.chomp` makes sense: if it wasn't part of the signature, then `'foo'.chomp(answer => 42)` would throw an `Unexpected named argument 'answer' passed` error, but it does not.  This is an important difference between the `Str.chomp` _method_ and the `chomp` _sub_; `chomp('foo', answer => 42)` does throw that error.

Given all that, I'd like to revise the `Str.chomp` method documentation to show that its signature includes `*%_`.

And everything that I said about `Str.chomp` applies just as much to **every** method that doesn't already document that it can take additional named arguments (some methods do already document this, for example by showing a capture argument).

In a comment in the PR, you said:

> In general, the policy here is to make definitions as close as possible, or the same, as they appear in the source. 

When you said "the source" did you mean "the source code of Rakudo"?  If so, I'm not sure that I agree that is (or should be) the policy.  My understanding is that our policy was to conform to Roast above all, which _often_ means matching Rakudo, but may not always.  Here, the signature – as defined by Roast – includes `*%_`.  In my opinion, the exact text of the Rakudo definition shouldn't matter.  But I'm new, and you have a **much** better sense of what our policies should be an and are.  If you feel that we shouldn't include `*%_` in our signature documentation, then I'll drop the matter and close this issue.

**Comment by codesections** [comment]
The OP mentioned that I discussed how to handle the implicit `*%_` on IRC before opening this issue.  I thought it might help to link to and summarize the relevant discussion.

* I was initially confused about why `.signature` showed a `*%_` parameter that, based on the docs, I wasn't expecting to see.  @lizmat helpfully explained that this is required due to Roast's notion of [Interface Consistency](https://github.com/Raku/roast/blob/master/S12-class/interface-consistency.t).  She also noted that there's no need to mention this in the documentation for each method.  [IRC log, 2020-08-21 at lines 227–2258](https://colabti.org/irclogger/irclogger_log/raku?date=2020-08-21#l227).
* The conversation died for a while, and then @jvc asked why we _shouldn't_ add `*%_` to the signatures that we list.  There was a bit of confusion about whether @jvc's suggestion was to add text to the documentation's narrative description of the method or just to the signature.  After @jvc clarified that they just meant adding the `*%_` to the signature, @lizmat said she would be fine with adding `*%_` to the signature. [IRC log, 2020-08-21 at lines 340–367](https://colabti.org/irclogger/irclogger_log/raku?date=2020-08-21#l340).
* @coke suggested adding an xtest to ensure that all the method signatures we documet have an invocant (which is another implicit item that is present in the `.signature` but not in the Rakudo source code documentation.  We _usually_ but not always include the invocant in our signature.  I'm currently implementing this potential xtest as a separate PR).  [IRC log, 2020-08-21 at line 372](https://colabti.org/irclogger/irclogger_log/raku?date=2020-08-21#l372).

**Comment by JJ** [comment]
> Again, my apologies for the miscommunication; I explained how it happened a bit in a comment #3601, and have updated this issue's title to clarify the scope of the proposed changes.
> 
> Since there was a miscommunication, let me back up and explain where I'm coming from with this change.
> 
>     * The documentation currently lists the signature for each documented method
> 
>     * Roast requires that `Method`s have a `signature` method, which also lists the signature for each method
> 
>     * In an **ideal** world (one with a 100% complete Roast, a 100% accurate implementation, and a 100% accurate set of docs), these two ways of viewing the signature for the method would show the same signature
> 
>     * In the real world, we'll likely never get to a point where the documented signatures and the signatures listed by Rakudo's `.signature` are exactly the same, but it's worth aiming for
> 
>     * Discrepancies between the documented signature and the `.signature` output could indicate a 1) a Rakudo bug, 2) an error in the docs, or 3) an under-specified area in Roast

But we need to strive to make the description of signatures in the documentation factually accurate, as well as actionable and useful, not only for the customer, but also whoever is working with it. The best functional reference we have is roast, the best reference we have for the signature is actually the source. *Not* using the source is not useful (because what's the point of listing useless arguments?) or actionable (why would you want to add useless arguments) and confusing for documentors (because they will need to check and check again if there was a good reason for them to be that way)

> 
> 
> Given all of the above, I would _like_ to get the signatures that we document closer to the signatures returned by `.signature`. Doing so would allow us to check for remaining discrepancies, which could indicate issues (either for us, Roast, or Rakudo). However, right now exactly **none** of the signatures we document are the same as the `.signature` signatures.

Unless there's a good reason for that, I don't think that's an objective we would aim for. And making a test pass is not a good reason. You can very easily chop the last argument of the signature and test again, and make it pass if that's the case, maybe with a `note`. 

> 
> Consider the `Str.chomp` method for example: If you consult [our docs](https://docs.raku.org/routine/chomp#(Str)_routine_chomp) you'll see that it has the signature
> 
> ```perl6fe
> (Str:D: --> Str:D)
> ```
> 
> However, if you run `Str.^lookup('chomp').signature` in a REPL, you'll get
> 
> ```perl6fe
> (Str:D: *%_ --> Str:D)
> ```
> 

OK, so what? Is that argument used elsewhere? In some cases (line `.lines`) it might be, but in general, it will not. Isn't it better to change the test than change the documentation? Also, these arguments will not be tested, in general. Are they not an implementation issue? If they are not, it will be in a case-by-case basis, and we'll need to address that. But *in general* I think it will simply be wrong.

> Additionally, this output from `.signature` is _not_ a choice Rakudo made – it's [explicitly specified in Roast](https://github.com/Raku/roast/blob/master/S12-class/interface-consistency.t#L19). And including `*%_` in the signature for `Str.chomp` makes sense: if it wasn't part of the signature, then `'foo'.chomp(answer => 42)` would throw an `Unexpected named argument 'answer' passed` error, but it doesn't. This is an important difference between the `Str.chomp` _method_ and the `chomp` _sub_; `chomp('foo', answer => 42)` does throw that error – but our current docs of the two signatures don't make this difference clear.
> 

You are using a single example for a general change of policy, as opposed to a note (which you can blow up as much as you want) in the `Method` page. Besides, in some case it will make sense; let's just change that.

> Given all that, I'd like to revise the `Str.chomp` method documentation to show that its signature includes `*%_`.

Please don't. Not going to repeat arguments here. 
> 
> And everything that I said about `Str.chomp` applies just as much to **every** method that doesn't already document that it can take additional named arguments (some methods do already document this, for example by showing a capture argument).
> 

If it applies to *every* method, document it in *Method*. Don't change the documents in every method. Does not make any sense to add an argument, if you're not going to say anything about it. And it does not make any sense to add "We have this argument, which is right there to not throw in case anyone uses named args" in *every* method.

> In a comment in the PR, you said:
> 
> > In general, the policy here is to make definitions as close as possible, or the same, as they appear in the source.
> 
> When you said "the source" did you mean "the source code of Rakudo"? If so, I'm not sure that I agree that is (or should be) the policy. My understanding is that our policy is to conform to Roast above all, which _often_ means matching Rakudo, but may not always. Here, the signature – as defined by Roast – includes `*%_`. In my opinion, the exact text of the Rakudo definition shouldn't matter. But I'm new, and you have a **much** better sense of what our policies should be an and are. If you feel that we shouldn't include `*%_` in our signature documentation, then I'll drop the matter and close this issue.

"is (or should be)" are two totally different things. That's the policy, because it's the most sensible one. You can try and figure out which definition fits the roast tests, but you can be sure that whatever is in raku *does* fit that. In some cases you might be tempted to simplify that for educational purposes (and we have done it).

> But I'm new, and you have a **much** better sense of what our policies should be an and are.

First, I really appreciate the work you've been doing and I really hope you will keep doing it. I don't know if I have a sense or I've totally lost it, so please look at my arguments before (and above) anything else. Baseline here is, let's add that argument where it really makes sense, that is, where those implicit arguments are placed through. It's really brilliant, and I really appreciate, that you've caught this mismatch between source and .signature returned. So let's use it where these arguments are really used, and only there. Since the general policy is "match the source" it might be helpful for future doccers (73 years from now) to maybe include a comment, or simply atomically change the definition and add that caveat as a commit message that can be seen when "git blame"d. Or a comment, or both, whatever looks the best. But let's please stick to the "match the source" policy, as much as we possibly can.

> If you feel that we shouldn't include `*%_` in our signature documentation, then I'll drop the matter and close this issue

I totally agree if you capitalize "signature". It should be included in Method, and possibly also in Signature, maybe additionally in method.signature, but not in every method. But please don't drop the matter, there are some valid changes in the PR and I'd love to incorporate them to the docs.

**Comment by JJ** [comment]
> The OP mentioned that I discussed how to handle the implicit `*%_` on IRC before opening this issue. I thought it might help to link to and summarize the relevant discussion.
> 
>     * I was initially confused about why `.signature` showed a `*%_` parameter that, based on the docs, I wasn't expecting to see.  @lizmat helpfully explained that this is required due to Roast's notion of [Interface Consistency](https://github.com/Raku/roast/blob/master/S12-class/interface-consistency.t).  She also noted that there's no need to mention this in the documentation for each method.  [IRC log, 2020-08-21 at lines 227–2258](https://colabti.org/irclogger/irclogger_log/raku?date=2020-08-21#l227).
> 

And you should do well *always* doing what @lizmat says. I really mean that.

>     * The conversation died for a while, and then @jvc asked why we _shouldn't_ add `*%_` to the signatures that we list.  There was a bit of confusion about whether @jvc's suggestion was to add text to the documentation's narrative description of the method or just to the signature.  After @jvc clarified that they just meant adding the `*%_` to the signature, @lizmat said she would be fine with adding `*%_` to the signature. [IRC log, 2020-08-21 at lines 340–367](https://colabti.org/irclogger/irclogger_log/raku?date=2020-08-21#l340).
> 

Well, er. She's saying she would be fine. Unfortunately, this is a change of policy which, of course, I'll have to accept if everyone else agrees with that. But for the time being I'd really prefer to keep it the way it is (with the caveats shown above).

>     * @coke suggested adding an xtest to ensure that all the method signatures we documet have an invocant (which is another implicit item that is present in the `.signature` but not in the Rakudo source code documentation.  We _usually_ but not always include the invocant in our signature.  I'm currently implementing this potential xtest as a separate PR).  [IRC log, 2020-08-21 at line 372](https://colabti.org/irclogger/irclogger_log/raku?date=2020-08-21#l372).

And that's a totally different thing, that was also included in the PR, meant just another change of policy, and should have been the subject of a different PR and a different discussion. Just for the record, it again breaks general policy, so I don't agree with that.

**Comment by codesections** [comment]
@JJ, thanks for the thoughtful reply; I believe I now understand your perspective on the issue a bit better.  Now that I do, I'll walk through the `Str.chomp` example again, and summarize your view (as I understand it) and my own.

* [Rakudo's source code](https://github.com/rakudo/rakudo/blob/master/src/core.c/Str.pm6#L110) declares `Str.chomp`  with the line `method chomp(Str:D: --> Str:D)`
This declaration does not include an explicit named slurpy but, due to the `method` keyword, it implicitly has one. 
* According to the `.signature` method, `Str.chomp`'s signature is `(Str:D: *%_ --> Str:D)`
* [Our documentation](https://docs.raku.org/routine/chomp#class_Str) currently follows the format from Rakudo source – it does not show the signature as reported by `.signature`.

**Your view**, as I understand it, is that it would be confusing to add `*%_` to the documented signature without an explanation because users might think that `*%_` had a function in the way `Str.chomp` is normally invoked (as opposed to being there to support `callsame` and friends).  You also think – and I agree – that we shouldn't add a note to `Str.chomp`'s description explaining `*%_`'s limited function; that would just be uselessly repetitive/verbose.  (Please feel free to correct me if I got your view wrong in any way).

**My view** is that it's confusing to document `Str.chomp` as having a different signature than the one returned by `.signature`.  I believe that new users rarely look at Rakudo's source code, but frequently use Raku's powerful introspection abilities, and it's confusing to have the docs tell them that `Str.chomp`'s signature is one thing but their REPL tell them (via `.signature`) that it's something else.  Seeing that inconsistency, users might think that the docs (or `.signature`) are buggy.

As I see it, we simply have a difference of opinion about what is more likely to be confusing to users.  We both want the docs to be as clear as possible, but disagree about the best way to achieve that goal.

## Plan

Since we simply disagree on this point, I'll defer to your judgment unless others chime in to say that they also feel that including `*%_` is significantly clearer than leaving it out.  I plan to leave this issue open for a few more days to give anyone the opportunity to speak up and then close it (assuming a consensus position that `*%_` should be added doesn't emerge).

> But let's please stick to the "match the source" policy, as much as we possibly can. … But please don't drop the matter, there are some valid changes in the PR and I'd love to incorporate them to the docs.

Given that, I'll split off the potentially-valid changes in this PR into separate issues/PRs that stick to the "match the source" policy.

**Comment by Altai-man** [comment]
IMO, documentation should describe API, not blood of some kinda obscure features.

Technically - yes, all methods include _implicit_ *%_ thingy, but consider this use-case (which is I am kind of sure dominating around): you are just an average Raku programmer. You want to know how to do thing X. You search for a routine or a class/object X which does the thing X. What you ultimately want to know are:

* What you have to give in
* What observable effects happen (in common and edge cases)
* What will be the exact result

That's all.

On Rakudo level, implementation level, there might be tons of multi candidates not specified anywhere, various tricks and other things, but it does not matter much, because it is the public API we should really care about (which is roast supposed to be?).

And if you are a curious programmer writing clever code and you need something like `nextsame`, you go to the page describing those things and there you can read that "all methods also include implicit *%_ so you can pass around things and do tricks".

If you are a clever programmer, you could go to e.g. "Traps" page, wondering mindlessly in order to maybe learn a new thing or two, and then read a section with maybe "If you see your named args are going funky, maybe this is what happens, be aware of this and read futher at (link)".

>I believe that new users rarely look at Rakudo's source code, but frequently use Raku's powerful introspection abilities

I dare to disagree. Life is life and there is a lot of mundane, trivial work around, much more than there is clever work, and it should work as well. If you are using MOP, here be dragons you can read about, it is not necessary for us mere mortals doing mundane things.

**Comment by codesections** [comment]
> IMO, documentation should describe API, not blood of some kinda obscure features.

> consider this use-case (which is I am kind of sure dominating around): you are just an average Raku programmer. 
> On Rakudo level, implementation level, there might be tons of multi candidates not specified anywhere, various tricks and other things, but it does not matter much, because it is the public API we should really care about (which is roast supposed to be?).

> Life is life and there is a lot of mundane, trivial work around, much more than there is clever work, and it should work as well. If you are using MOP, here be dragons you can read about, it is not necessary for us mere mortals doing mundane things.

I agree with that 100% (or more, if we'll allow such things!)  The docs should **absolutely** be written to clarify the mundane aspects of the language and not the more advanced/meta features.  I was definitely thinking about things from that perspective when I proposed this change – or at least trying to!

Here's a view into my workflow, as someone who is still pretty new to Raku: I have one or more Raku files open, and a Raku REPL running.  I frequently pop over to the REPL to try out simple code or use introspection to help figure out how what I can do.  So, if I didn't remember that `Str.trim-trailing` exists, I might have a REPL session that looks like this (with the comments representing my thoughts)

```sh
> # hmm, was it a called trim-right?
> ' foo '.trim-right    
No such method 'trim-right' for invocant of type 'Str'
  in block <unit> at <unknown file> line 1
> # nope.  Maybe it was a named argument to .trim?
> dd ' foo '.trim(:right)
"foo"
Nil
> # Well, that didn't work.  Is that the wrong name for the argument?
> #  What arguments does trim take, anyway?
> Str.^lookup('trim').signature
(Str:D: *%_ --> Str:D)
> # OK, I guess it's not the trim method.
> Str.^methods(:local)
(BUILD Capture Int Num Version chomp pred succ match subst-mutate parse-base samecase samemark samespace word-by-word trim-leading trim-trailing trim encode NFC NFD NFKC NFKD unival univals wordcase trans parse-n\
ames uniparse indent codes chars uc lc tc fc tclc flip ord Date DateTime IO WHY WHICH Bool Str Stringy DUMP ACCEPTS chop starts-with ends-with substr-eq contains indices index rindex Numeric gist raku comb subst \
ords lines split words substr substr-rw BUILDALL)
> # Oooh, trim-trailing!
> ' foo '.trim-trailing
" foo"
Nil
> # *there* we go
```

That's a bit of a contrived example, but my point is that I spend a lot of time in the REPL and use Raku's introspection abilities to explore the language without the context-shift inherent in switching to reading the documentation.  Of course, often just knowing the signature of a method isn't enough to know how to use it correct, so I _do_ switch over to reading the docs.  When I do, I've been confused by why the signature I was looking at in the REPL isn't the same as the one I see in the docs – which is what got me started on all this in the first place.  I got confused, asked on IRC about the difference, and then thought other new users might be confused too.

From the way you talked about introspecting with `.signature` ("MOP", "blood of some kinda obscure features", "here be dragons") it sounds like your workflow isn't all that similar to mine.  I'd be interested in hearing more about it.  Maybe my workflow just isn't than common, and I'm trying to address a point of confusion that came up for **me** but that wouldn't come up for others.

**Comment by Altai-man** [comment]
>use Raku's introspection abilities to explore the language without the context-shift inherent

While people are free to use tools they want, I believe that if someone uses an advanced enough emacs/vim/whatever or an IDE, in your example there are two points: 1)`.trim-trailing` will be auto-completed for you; 2)Documentation (quite likely taken from this docs processed) will be shown if you ask for it, inlined, no switches needed.

I may be wrong, but I'd say a lack of tooling (features) is breaking the workflow, not docs signatures. Also, auto-completion will show you parameters and candidates as well.

>"MOP"

Meta-Object Protocol is just a term which is not strictly related to existence of Signature class, but I usually tend to lump it into a general pack of things which I label with "Advanced" stamp, including first-class support for Signatures or nextsame business or augmenting language slangs.

>"blood of some kinda obscure features"

The word "guts" in meaning "internals" is probably more fitting, my non-native English has failed me for a bit. Yet, I do not consider calling what I described above "obscure" as something wrong. If there are negative connotations, they were not implied. Let's stick to the word "advanced" instead.

>here be dragons

Warnings like those are common when it comes to advanced topics, I think. We have one, for example, at https://docs.raku.org/language/mop#Power_and_responsibility

An admin writing a maintenance script highly likely wants to use signatures and trim strings, but there is hardly a room to introspect signatures for doing a $dayjob.
A library writer sometimes may want to use signature introspection when writing something which can do different things depending on what is provided. For example, take Cro router DSL:

```
route {
    get -> 'foo', 'bar', Int $id { #`[ do something when url "foo/bar/$id" is visited ] }
}
```
Here, `get` routine takes a block of code with a signature and introspects it so that HTTP requests with particular URL were routed to it if it matches.

How often do people write DSLs? :>

Now about the workflow thing. Mine is nothing special or advanced. When I wrote various bits of Cro (learning Raku) or did some personal projects, I wrote code in emacs and later in Comma. Before that I read a couple of intros to the language, some tutorials. Let's say I want to find `.trim-trailing`. Go to docs web page. Find class `Str`, go there, browser search, `trim` there, oh, we see something. There is a signature now, takes a string, returns string, okay, the description is okay.

I would add a simple example maybe, it wouldn't hurt. That's about it.

Also, I would be really delighted if we could do some work towards the solution for, erm. Sigh. So there was a ticket and now I couldn't find it, unfortunately, maybe someone who is brave enough to read all of my long ranting can find it. The ticket was suggesting to improve the website so that the user can hover mouse over a signature part and a tooltip will be shown with an automatically generated explanation for this bit (or whole signature, doesn't matter much and maybe depends on how to implement). Similar to "regex explainers" around.

For example, if you take signature of trim-trailing, `method trim-trailing(Str:D: --> Str)`, for a novice it could not be obvious how `Str:D:` (with a colon at the end) means "self", and otherwise it would mean "Str in - Str out", which is wrong, so a tooltip generated to be `It is called on a Defined Str instance and returns a Str` would be really de-confusing for novices.

But *%_ (in just my opinion, of course) should be described in a couple of places where it makes sense for people to look up for clever things and not be present just everywhere. Without a quick and easy explanation of what *%_ is, why is it so and such, additional things which appear magical (I mean, it's Raku, but to someone else it is technically "an asterisk, percent sign and underscore" in a row) would only confuse more, not less.
